src/engine-actions/get-payment-preimage.js

```diff
@@ -27,62 +20,82 @@ const POLLING_INTERVAL = 10000
  * then return the hash.
  *
  * @param  {string} paymentHash - Base64 encoded payment hash
- * @returns {GetPaymentPreimageOutcome}
+ * @returns {Promise<GetPaymentPreimageOutcome>}
  * @throws {Error} If payment is in an unknown status
  */
 async function getPaymentPreimage (paymentHash) {
-  const { client, logger } = this
-  const { status } = await lookupPaymentStatus(paymentHash, { client })
+  return new Promise(async (resolve, reject) => {
+    try {
+      if (!paymentHash) {
+        throw new Error('paymentHash must be defined')
+      }
 
-  if (status === PAYMENT_STATUSES.GROUNDED) {
-    return { permanentError: `No payment with hash '${paymentHash}' is in-flight or complete` }
-  }
+      // TODO: look into how gRPC wants `bytes` in JS
+      const paymentHashBuf = Buffer.from(paymentHash, 'base64')
 
-  if (status === PAYMENT_STATUSES.COMPLETED) {
-    return { paymentPreimage: await getCompletedPreimage(paymentHash, { logger, client }) }
-  }
+      this.logger.debug('Attempting to track payment for paymentHash:', { paymentHash })
 
-  if (status === PAYMENT_STATUSES.IN_FLIGHT) {
-    // if the payment is still in-flight, we'll wait and try again later
-    await delay(POLLING_INTERVAL)
-    return getPaymentPreimage.call(this, paymentHash)
-  }
+      const stream = await trackPayment(paymentHashBuf, { client: this.client })
 
-  throw new Error(`Unknown payment status ${status} for payment with hash '${paymentHash}'`)
-}
+      const errorListener = (err) => {
+        // TODO: handle gRPC NOT FOUND errors
+        //  It may be best to handle it as a `permanentError`, so we don't have to
+        //  rely on checking NOT FOUND in downstream logic
+        this.logger.error(`Error from trackPayment stream`)
```

Comment:
Nice to have identifying information in this logs, you can pass an object at the end to add info.